### PR TITLE
feat: Support '"' quoting for custom signing tool

### PR DESCRIPTION
**Pull request by anatawa12**
Add support for '"' quoting in custom signing tool

https://github.com/tauri-apps/tauri/issues/10583#issuecomment-2285266577

### CODE REVIEW

File: tooling/bundler/src/bundle/windows/sign.rs [hunk 1]
```diff
@@ -132,10 +132,95 @@ pub fn verify(path: &Path) -> crate::Result<bool> {
...
+  let mut state = State::Idle;
+
+  for c in command.chars() {
+    match (state, c) {
+      (State::Idle, c) if c.is_ascii_whitespace() => {
+        // skip whitespace
+      }
+      (State::Idle, '"') => {
+        state = State::Quoted;
+      }
```
Comment (anatawa12) [inline tooling/bundler/src/bundle/windows/sign.rs:163]: It might be good to reserve some symbols for future extension

**Comment by anatawa12** [comment]
I've fixed format error. sorry

**Review by lucasfernog — APPROVED** [review]
looks good to me, anything against this change @amrbashir ?

File: tooling/bundler/src/bundle/windows/sign.rs [hunk 2]
```diff
@@ -132,10 +132,95 @@ pub fn verify(path: &Path) -> crate::Result<bool> {
...
+        state = State::Unquoted;
+      }
+
+      (State::Unquoted, c) if c.is_ascii_whitespace() => {
+        result.push(buffer.clone());
+        buffer.clear();
+        state = State::Idle;
+      }
+      (State::Unquoted, '"') => {
+        return Err(anyhow::anyhow!("custom signing command contains an unclosed quote").into());
```
Comment (amrbashir) [inline tooling/bundler/src/bundle/windows/sign.rs:175]: We shouldn't error here imo, and just continue parsing

File: tooling/bundler/src/bundle/windows/sign.rs [hunk 3]
```diff
@@ -132,10 +132,95 @@ pub fn verify(path: &Path) -> crate::Result<bool> {
...
+        buffer.push('"');
+        state = State::Quoted;
+      }
+      (State::QuotedJustAfterQuote, c) if c.is_ascii_whitespace() => {
+        result.push(buffer.clone());
+        buffer.clear();
+        state = State::Idle;
+      }
+      (State::QuotedJustAfterQuote, _) => {
+        return Err(anyhow::anyhow!("custom signing command contains an unclosed quote").into());
```
Comment (amrbashir) [inline tooling/bundler/src/bundle/windows/sign.rs:198]: same here

File: tooling/bundler/src/bundle/windows/sign.rs [hunk 4]
```diff
@@ -132,10 +132,95 @@ pub fn verify(path: &Path) -> crate::Result<bool> {
...
+    }
+  }
+
+  match state {
+    State::Idle => {}
+    State::Unquoted => {
+      result.push(buffer);
+    }
+    State::Quoted => {
+      return Err(anyhow::anyhow!("custom signing command contains an unclosed quote").into());
```
Comment (amrbashir) [inline tooling/bundler/src/bundle/windows/sign.rs:209]: and here

**Comment by amrbashir** [comment]
I wonder if it is better to change the command into an object rather than introduce custom parsing logic
```json
{
	"customSignCommand": {
		"cmd": "program with space.exe",
		"args": ["list", "of", "args"],
	}
}
```

File: tooling/bundler/src/bundle/windows/sign.rs (hunk 2 above)
Comment (anatawa12) [reply tooling/bundler/src/bundle/windows/sign.rs:175]: I think it's a little unclear that `foo"bar baz` should be treated as `foo` `bar baz`, `foo"bar` `baz`, or `foo"bar baz` so I made this error. Which should I choose?

**Comment by anatawa12** [comment]
I think it's better to use object form but I have no Idea how should I implement that since crate for bundler and config is diffdrent crate Shuld I create struct on both config and bundler?

File: tooling/bundler/src/bundle/windows/sign.rs (hunk 2 above)
Comment (amrbashir) [reply tooling/bundler/src/bundle/windows/sign.rs:175]: it should be `foo"bar` `baz`

**Comment by amrbashir** [comment]
> I think it's better to use object form but I have no Idea how should I implement that since crate for bundler and config is diffdrent crate Shuld I create struct on both config and bundler?

yup, two configs, there is similar objects you can take inspiration from like `NsisConfig` in config and `NsisSettings` in bundler.

File: tooling/bundler/src/bundle/windows/sign.rs (hunk 2 above)
Comment (amrbashir) [reply tooling/bundler/src/bundle/windows/sign.rs:175]: but if we go with the object solution (which we should), this doesn't matter

Comment (anatawa12) [reply tooling/bundler/src/bundle/windows/sign.rs:175]: I mean your decision is not clear and reasonable enough I think.
For example, with command prompt (`cmd.exe`), `foo"bar baz` become `foobar baz`.


(Sorry I forget to list this up in the previous example list. `foo"bar baz` is mistake of `foobar baz`)


Anyway I made PR for object solution so this discussion might be useless.

#10634

Comment (amrbashir) [reply tooling/bundler/src/bundle/windows/sign.rs:175]: thanks, will close this PR then and move discussion there

**Pull request closed without merging**